translationUtils.ts

```diff
@@ -0,0 +1,284 @@
+				// It is not an exceptional event, so simply return null.
+				resolve(null);
+			})
+	);
+}
+
+// Returns null if url is malformed.
+export function tryParseUrl(url: string): URL | null {
+	try {
+		return new URL(url);
+	} catch {
+		return null;
+	}
+}
+
+// Returns null if the image cannot be translated.
+export async function translate(image: ImageBase64): Promise<ImageBase64 | null> {
+	const translationRequest = await fetch('http://localhost:3001/translate', {
+		method: 'POST',
+		headers: {
+			'Content-Type': 'application/json'
+		},
+		// API supports multiple images, but only one is necessary for now.
+		body: JSON.stringify({ base64Images: [image], apiKey: '[REDACTED]' })
+	});
+
+	if (translationRequest.status !== StatusCode.Ok) {
+		throw new Error(`translation request failed with:\n${JSON.stringify(translationRequest)}`);
+	}
+
+	const translations = (await translationRequest.json()).images[0] as TranslationResult[];
+	const isAllSupported = translations.every(checkIsSupportedLanguage);
+	if (translations.length === 0 || !isAllSupported) {
+		return null;
+	}
+
+	const translatedImage = await overlayTranslations(image, translations);
+	return translatedImage;
+}
+
+// Returns null if the base64 data could not be retrieved.
+export function getBase64Data(
+	imageUrl: URL
+): Promise<{ base64Data: ImageBase64; width: number; height: number } | null> {
+	return new Promise(resolve => {
+		const image = new Image();
+		image.onload = function () {
+			const canvas = createCanvas(image.width, image.height);
+			const context = canvas.getContext('2d');
+			context.drawImage(
+				image,
+				0,
+				0,
+				image.width,
+				image.height,
+				0,
+				0,
+				image.width,
+				image.height
+			);
+			resolve({
+				base64Data: canvas.toDataURL('image/png'),
+				height: image.height,
+				width: image.width
+			});
+		};
+
+		// Handle various things that can go wrong when loading an image.
+		// This includes network issues, dead links, etc.
+		function onFailed() {
+			resolve(null);
+		}
+		image.onerror = onFailed;
+
+		image.src = imageUrl.toString();
+	});
+}
+
+function checkIsSupportedLanguage(translation: TranslationResult) {
+	const supportedLanguages = ['jp', 'kr', 'cn'];
+	return supportedLanguages.includes(translation.originalLanguage);
+}
+
+function overlayTranslations(
+	imageBase64: string,
+	translations: TranslationResult[]
+): Promise<ImageBase64> {
+	return new Promise(resolve => {
+		const image = new Image();
+		image.onload = function () {
+			const canvas = createCanvas(image.width, image.height);
+			const context = canvas.getContext('2d');
+			context.drawImage(
+				image,
+				0,
+				0,
+				image.width,
+				image.height,
+				0,
+				0,
+				image.width,
+				image.height
+			);
+
+			const filteredTranslations = translations.filter(x => x.translatedText.trim() !== '');
+
+			// Draw text bubble as a rounded rectangle.
+			for (const translation of filteredTranslations) {
```

Comment:
This for loop and the one below are identical. This is not a mistake. Again like with `if` statements- don't merge for loops together just because you can!

Two for loops might be easier to reason about (or it might be worse). Loops are already a more confusing control flow. Keep them as simple as possible by only doing one thing. Obviously performance concerns might tilt the scales (or other reasons), but don't assume performance until you've actually measured.

In this scenario, I cannot merge them anyways - I need to draw the boxes first so that they don't accidentally hide text.